File: C++/Sudoku.cpp
#include <iostream>
#include <array>
#include <deque>
#include <vector>
using namespace std;

class Sudoku {

    public:

        static string solve(string input) {
            Sudoku s;
            try {
                return s.read(input).search().show();
            }
            catch (NoSolutionException) {
                return "NO SOLUTION";
            }
        }

        static void init() {
            parents.resize(COORD);
            for (int i = 0; i < SIZE; i++) {
                for (int j = 0; j < SIZE; j++) {
                    int p = i / UNIT * UNIT + j / UNIT;
                    for (int k = 0; k < SIZE; k++) {
                        vector<Block>& vec = parents[coord(i, j, k)];
                        vec.push_back(block(GRD, i, j));
                        vec.push_back(block(ROW, i, k));
                        vec.push_back(block(COL, j, k));
                        vec.push_back(block(BOX, p, k));
                    }
                }
            }
            children.resize(BLOCK);
            for (Coord c = 0; c < COORD; c++)
                for (Block b : parents[c])
                    children[b].push_back(c);
        }

    private:

        enum View { GRD, ROW, COL, BOX };
        enum State { OPEN, FIXED, BANNED };

        typedef int Coord;
        typedef int Block;
        typedef int Count;

        static const int UNIT = 3;
        static const int SIZE = UNIT * UNIT;
        static const int COORD = SIZE * SIZE * SIZE;
        static const int BLOCK = 4 * SIZE * SIZE;
        static const Count DONE = SIZE + 1;
        static const string DIGITS;

        static vector<vector<Block>> parents;
        static vector<vector<Coord>> children;

        static class NoSolutionException {} NO_SOLUTION_EXCEPTION;

        static Coord coord(int i, int j, int k) {
            return (i * SIZE + j) * SIZE + k;
        }

        static Block block(View v, int p, int q) {
            return (v * SIZE + p) * SIZE + q;
        }

        array<State, COORD> state;
        array<Count, BLOCK> count;
        deque<Coord> queue;

        Sudoku() {
            state.fill(OPEN);
            count.fill(SIZE);
        }

        Sudoku(const Sudoku& that) {
            state = that.state;
            count = that.count;
        }

        Sudoku& enqueue(Coord c) {
            queue.push_back(c);
            return *this;
        }

        Sudoku& read(string input) {
            int len = input.length();
            for (int i = 0; i < SIZE; i++) {
                for (int j = 0; j < SIZE; j++) {
                    if (i * SIZE + j < len) {
                        int k = DIGITS.find(input[i * SIZE + j]);
                        if (k != -1)
                            enqueue(coord(i, j, k));
                    }
                }
            }
            return *this;
        }

        string show() {
            string result(SIZE * SIZE, '.');
            for (int i = 0; i < SIZE; i++)
                for (int j = 0; j < SIZE; j++)
                    for (int k = 0; k < SIZE; k++)
                        if (state[coord(i, j, k)] == FIXED)
                            result[i * SIZE + j] = DIGITS[k];
            return result;
        }

        Sudoku search() {

            while (not queue.empty()) {
                Coord c = queue.front();
                queue.pop_front();
                if (state[c] == FIXED)
                    continue;
                if (state[c] == BANNED)
                    throw NO_SOLUTION_EXCEPTION;
                fix(c);
            }

            Count min = DONE;
            Block bmin;
            for (Block b = 0; b < BLOCK; b++) {
                if (count[b] < min) {
                    min = count[b];
                    bmin = b;
                }
            }

            if (min == DONE)
                return *this;

            for (Coord c : children[bmin]) {
                if (state[c] == BANNED)
                    continue;
                Sudoku s(*this);
                try {
                    return s.enqueue(c).search();
                }
                catch (NoSolutionException) {
                }
            }

            throw NO_SOLUTION_EXCEPTION;
        }

        void fix(Coord c) {
            state[c] = FIXED;
            for (Block b : parents[c])
                mark_done(b);
        }

        void mark_done(Block b) {
            count[b] = DONE;
            for (Coord c : children[b])
                if (state[c] == OPEN)
                    ban(c);
        }

        void ban(Coord c) {
            state[c] = BANNED;
            for (Block b : parents[c])
                if (count[b] != DONE)
                    countdown(b);
        }

        void countdown(Block b) {
            count[b]--;
            if (count[b] == 0)
                throw NO_SOLUTION_EXCEPTION;
            if (count[b] == 1)
                for (Coord c : children[b])
                    if (state[c] == OPEN)
                        enqueue(c);
        }
};

const string Sudoku::DIGITS = "123456789";

vector<vector<Sudoku::Block>> Sudoku::parents;
vector<vector<Sudoku::Coord>> Sudoku::children;

int main() {
    Sudoku::init();
    string input;
    while (cin >> input)
        cout << Sudoku::solve(input) << endl;
}

File: Kotlin/Sudoku.kt
fun main(args: Array<String>) {
    while (true) {
        println(Sudoku.solve(readLine() ?: return))
    }
}

class Sudoku {

    enum class View { GRD, ROW, COL, BOX }

    enum class State { OPEN, FIXED, BANNED }

    class NoChoice : Exception("NoChoice");

    companion object {

        const val UNIT  = 3
        const val SIZE  = UNIT * UNIT
        const val COORD = SIZE * SIZE * SIZE
        const val BLOCK = 4 * SIZE * SIZE
        const val DONE  = SIZE + 1

        const val DIGITS = "123456789"

        fun coord(i: Int, j: Int, k: Int): Int = (i * SIZE + j) * SIZE + k

        fun block(v: View, p: Int, q: Int): Int = when (v) {
            View.GRD -> (0 * SIZE + p) * SIZE + q
            View.ROW -> (1 * SIZE + p) * SIZE + q
            View.COL -> (2 * SIZE + p) * SIZE + q
            View.BOX -> (3 * SIZE + p) * SIZE + q
        }

        val parents = Array(COORD) { mutableListOf<Int>() } .apply {
            for (i in 0 until SIZE) {
                for (j in 0 until SIZE) {
                    val p = i / UNIT * UNIT + j / UNIT
                    for (k in 0 until SIZE) {
                        get(coord(i, j, k)).apply {
                            add(block(View.GRD, i, j))
                            add(block(View.ROW, i, k))
                            add(block(View.COL, j, k))
                            add(block(View.BOX, p, k))
                        }
                    }
                }
            }
        }

        val children = Array(BLOCK) { mutableListOf<Int>() } .apply {
            for (c in 0 until COORD)
                for (b in parents[c])
                    get(b).add(c)
        }

        fun solve(input: String): String {
            return Sudoku().read(input).search()?.show() ?: "NO SOLUTION"
        }
    }

    val state: Array<State>
    val count: IntArray
    val queue: MutableList<Int>

    constructor() {
        state = Array<State>(COORD) { State.OPEN }
        count = IntArray(BLOCK) { SIZE }
        queue = mutableListOf<Int>()
    }

    constructor(that: Sudoku) {
        state = that.state.clone()
        count = that.count.clone()
        queue = mutableListOf<Int>()
    }

    fun enqueue(c: Int): Sudoku {
        queue.add(c)
        return this
    }

    fun read(input: String): Sudoku {
        val len = input.length
        for (i in 0 until SIZE) {
            for (j in 0 until SIZE) {
                if (i * SIZE + j < len) {
                    val k = DIGITS.indexOf(input[i * SIZE + j])
                    if (k != -1)
                        enqueue(coord(i, j, k))
                }
            }
        }
        return this
    }

    fun show(): String {
        val output = CharArray(SIZE * SIZE) { '.' }
        for (i in 0 until SIZE)
            for (j in 0 until SIZE)
                for (k in 0 until SIZE)
                    if (state[coord(i, j, k)] == State.FIXED)
                        output[i * SIZE + j] = DIGITS[k]
        return String(output)
    }

    fun search(): Sudoku? {

        while (!queue.isEmpty()) {
            val c = queue.removeAt(0)
            when (state[c]) {
                State.OPEN ->
                    try { fix(c) } catch (_: NoChoice) { return null }
                State.FIXED ->
                    Unit
                State.BANNED ->
                    return null
            }
        }

        val m = count.min()!!
        if (m == DONE)
            return this

        val b = count.indexOf(m)
        for (c in children[b])
            if (state[c] == State.OPEN)
                Sudoku(this).enqueue(c).search()?.let { return it }

        return null
    }

    fun fix(c: Int) {
        state[c] = State.FIXED
        for (b in parents[c])
            mark(b)
    }

    fun mark(b: Int) {
        count[b] = DONE
        for (c in children[b])
            if (state[c] == State.OPEN)
                ban(c)
    }

    fun ban(c: Int) {
        state[c] = State.BANNED
        for (b in parents[c])
            if (count[b] != DONE)
                countdown(b)
    }

    fun countdown(b: Int) {
        count[b]--
        when (count[b]) {
            0 -> throw NoChoice();
            1 ->
                for (c in children[b])
                    if (state[c] == State.OPEN)
                        enqueue(c);
            else -> Unit
        }
    }
}
